File: spec/java/src/io/kaitai/struct/spec/TestSwitchManualEnum.java
package io.kaitai.struct.spec;

import io.kaitai.struct.testformats.SwitchManualEnum;
import io.kaitai.struct.testformats.SwitchManualEnum.Opcode.*;

import org.testng.annotations.Test;

import static org.testng.Assert.assertEquals;

public class TestSwitchManualEnum extends CommonSpec {
    @Test
    public void testSwitchManualEnum() throws Exception {
        SwitchManualEnum r = SwitchManualEnum.fromFile(SRC_DIR + "switch_opcodes.bin");

        assertEquals(r.opcodes().size(), 4);

        assertEquals(r.opcodes().get(0).code(), Code.STRVAL);
        assertEquals(((Strval) r.opcodes().get(0).body()).value(), "foobar");

        assertEquals(r.opcodes().get(1).code(), Code.INTVAL);
        assertEquals(((Intval) r.opcodes().get(1).body()).value(), 0x42);

        assertEquals(r.opcodes().get(2).code(), Code.INTVAL);
        assertEquals(((Intval) r.opcodes().get(2).body()).value(), 0x37);

        assertEquals(r.opcodes().get(3).code(), Code.STRVAL);
        assertEquals(((Strval) r.opcodes().get(3).body()).value(), "");
    }
}

File: spec/cpp_stl_11/test_switch_manual_enum.cpp
#include <boost/test/unit_test.hpp>

#include <switch_manual_enum.h>

#include <iostream>
#include <fstream>
#include <vector>

BOOST_AUTO_TEST_CASE(test_switch_manual_enum) {
    std::ifstream ifs("src/switch_opcodes.bin", std::ifstream::binary);
    kaitai::kstream ks(&ifs);
    switch_manual_enum_t* r = new switch_manual_enum_t(&ks);

    BOOST_CHECK_EQUAL(r->opcodes()->size(), 4);

    BOOST_CHECK_EQUAL(r->opcodes()->at(0)->code(), switch_manual_enum_t::opcode_t::CODE_ENUM_STRVAL);
    BOOST_CHECK_EQUAL((static_cast<switch_manual_enum_t::opcode_t::strval_t*>(r->opcodes()->at(0)->body()))->value(), "foobar");

    BOOST_CHECK_EQUAL(r->opcodes()->at(1)->code(), switch_manual_enum_t::opcode_t::CODE_ENUM_INTVAL);
    BOOST_CHECK_EQUAL((static_cast<switch_manual_enum_t::opcode_t::intval_t*>(r->opcodes()->at(1)->body()))->value(), 0x42);

    BOOST_CHECK_EQUAL(r->opcodes()->at(2)->code(), switch_manual_enum_t::opcode_t::CODE_ENUM_INTVAL);
    BOOST_CHECK_EQUAL((static_cast<switch_manual_enum_t::opcode_t::intval_t*>(r->opcodes()->at(2)->body()))->value(), 0x37);

    BOOST_CHECK_EQUAL(r->opcodes()->at(3)->code(), switch_manual_enum_t::opcode_t::CODE_ENUM_STRVAL);
    BOOST_CHECK_EQUAL((static_cast<switch_manual_enum_t::opcode_t::strval_t*>(r->opcodes()->at(3)->body()))->value(), "");

    delete r;
}
